src/rade_api.h

```diff
@@ -96,11 +96,16 @@ RADE_EXPORT int rade_n_tx_out(struct rade *r);
 RADE_EXPORT int rade_n_tx_eoo_out(struct rade *r);
 RADE_EXPORT int rade_nin_max(struct rade *r);
 RADE_EXPORT int rade_n_features_in_out(struct rade *r);
+RADE_EXPORT int rade_n_eoo_bits(struct rade *r);
 
 // note vocoder is not encapsulated in API in this version
 // returns number of RADE_COMP samples written to tx_out[]
 RADE_EXPORT int rade_tx(struct rade *r, RADE_COMP tx_out[], float features_in[]);
 
+// Set the rade_n_eoo_bits() bits to be sent in the EOO frame, which are
+// in +/- 1 float form (note NOT 1 or 0)
+RADE_EXPORT void rade_tx_set_eoo_bits(struct rade *r, float eoo_bits[]);
```

Comment:
It seems that the code from codec2 just passes the bits directly for TX instead of using floats. Is that viable to do here too? If not, is there already some sort of way to convert to the form this function needs?

Reply:
Can you pls describe the format of the bits generated by reliable text?

Reply:
Basically, the existing codec2 reliable_text system ties into the varicode functionality. When you set a callsign (https://github.com/drowe67/codec2/blob/main/src/reliable_text.c#L384), it does the following:

1. Converts the ASCII callsign into a special six-bit character set. This effectively generates a 48-bit string of zeros and ones.
2. Calculates the CRC8 of the above 48-bit string and appends it to the end. This results in 56 bits of data.
3. Generates parity bits with the `HRA_56_56` LDPC code and appends to the end of (2). (I have mentions of LDPC(112,56) in the comments but that might be wrong.) We now have 112 bits of data.
4. Interleaves the parity and data bits together.
5. Saves this result somewhere in memory.

When codec2 calls the TX callback that reliable_text injects (https://github.com/drowe67/codec2/blob/main/src/reliable_text.c#L340), the latter returns the next byte of data from (5), i.e.

* Call 1: `return txString[0]`
* Call 2: `return txString[1]`
* ...
* Call 14: `return txString[13]`
* Call 15: `return txString[0]`
* ...

This is then injected into the data transmitted over the air. 

Anyway, I hope this helps. Let me know if you need further clarification.

Reply:
OK so I think this is the array of bits to be sent to the modem:
```
char tx_text[LDPC_TOTAL_SIZE_BITS + RELIABLE_TEXT_UW_LENGTH_BITS];
```
So the format is one bit per char.  We'll need some glue code to connect reliable text to the EOO modem. So on the tx side we want a binary 1 -> +1, and binary 0 -> -1, something like:
```
eoo_bit[i] = 2.0*tx_text[i] - 1.0;
```
The LDPC decoder expects the received symbol to be mapped to the QPSK constellation defined by (`codec2/src/mp_decode.c`:
```
static COMP S_matrix[] = {
    {1.0f, 0.0f}, {0.0f, 1.0f}, {0.0f, -1.0f}, {-1.0f, 0.0f}};
```
which is 45 degrees rotated from the EOO text constellation {1,1}, {-1,1}, {0-1,-1}, {1,-1} .  So on the rx side, something like (ROT45 from `codec2/src/ofdm.c`):
```
complex float symbol = (COMP)eoo_out[2*i];
reliable_text_demod_symbol[i] = symbol * cmplx(ROT45);
```


Good idea to have a unit tests that runs with no noise, make sure the LDPC decoding happens with 1 iteration, just to make sure we haven't messed up anything.  Then add some noise, e.g. 2-3dB AWGN and make sure it hangs on.  Not sure where to do the tests .. it's a mash up of several repos :thinking:

Reply:
> Not sure where to do the tests .. it's a mash up of several repos

Considering the current API, I wonder if having this in freedv-gui would be best. That way, it'd have access to some of the codec2 functions (e.g. LDPC and interleaving) and we wouldn't need to reimplement versions in this repo. OTOH testing does become a bit harder but maybe doable with the current test framework. If that sounds good, I'll go ahead and do that.

Reply:
Yes I agree, it's the only project with everything linked in.  And you now have a nice command line test framework for `freedv-gui` to support automated tests :+1

Reply:
OK, I created https://github.com/drowe67/freedv-gui/pull/783 but I suspect there's something wrong with either the TX or RX symbol generation; BER is much too high when playing back files generated locally (i.e. no RF). I also disabled interleaving just to be sure that wasn't interfering. I'll have to investigate more when I have time.